[Python] Flaky test postgres `test_device_enrollment[device]`

The job https://github.com/Scille/parsec-cloud/actions/runs/8874398577/job/24361773207 fail at ` tests/api_v4/test_enrollment.py:153` with `AssertionError`:

```python
 assert Ok { last: true, payload: b"<from greeter 1>" } == Ok { last: false, payload: b"<from greeter 1>" }
```



<details>
<summary>Details</summary>

```python
    async def claimer_step_4() -> None:
        rep = await invited_rpc.invite_4_claimer_communicate(
            payload=b"<from claimer 1>",
        )
>       assert rep == invited_cmds.v4.invite_4_claimer_communicate.RepOk(
            payload=b"<from greeter 1>",
            last=False,
        )
E       assert Ok { last: true, payload: b"<from greeter 1>" } == Ok { last: false, payload: b"<from greeter 1>" }
E        +  where Ok { last: false, payload: b"<from greeter 1>" } = <class 'parsec._parsec.invited_cmds.v4.invite_4_claimer_communicate.RepOk'>(payload=b'<from greeter 1>', last=False)
E        +    where <class 'parsec._parsec.invited_cmds.v4.invite_4_claimer_communicate.RepOk'> = <module 'invite_4_claimer_communicate'>.RepOk
E        +      where <module 'invite_4_claimer_communicate'> = <module 'v4'>.invite_4_claimer_communicate
E        +        where <module 'v4'> = invited_cmds.v4
```

Stdout:

```txt
Step 1
Step 2
Step greeter 2b
Step claimer 2b
Step 3
Step greeter 3b
Step claimer 3b
Step 4
Step greeter 4 (last communicate)
```

</details>

<details>
<summary>Logs</summary>

```txt
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('6247faa8-a02e-407e-bffa-80b701ffb96f'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('f4e26c4b-1b02-4e4b-be80-f324c51df7f6'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventInvitation(event_id=UUID('07ebd9d8-3e52-4a47-ab37-b01b63db690b'), type='INVITATION', organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"), status=Ready)
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('f529279e-7718-4423-a78a-22dcac8dcfbc'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/authenticated/Org1 "HTTP/1.1 200 OK"
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/invited/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventInvitation(event_id=UUID('fc009139-1b17-45a1-ad02-c385c54152a1'), type='INVITATION', organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"), status=Idle)
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('147ed90d-6c22-4ef7-a893-8ea5252534ab'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('37a4099d-b475-4454-99b4-385b0c476cd8'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('c41cce9a-ade6-4eea-bb4b-7f2a2706beb1'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/authenticated/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('b36835a3-1497-4cc2-812c-49878e216357'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('34742eed-afed-415d-836a-3163e00c28b5'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('82831640-dcf8-4006-bd89-3085772ac648'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/invited/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('8cbf1061-211e-4eef-9130-0114f54dac90'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('758b48ea-2ca0-4f44-935f-a08278c44835'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('28063955-f1c0-436f-86b5-c2a8a9265ac8'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/authenticated/Org1 "HTTP/1.1 200 OK"
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/invited/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('e1693218-9de0-4c33-9399-884466b8d79e'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('0ce501ae-9d84-48e6-b50e-1b1131d82879'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('45a7c8b6-9e4f-4f09-af88-b8ee304d9d9e'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/authenticated/Org1 "HTTP/1.1 200 OK"
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/invited/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('3d5049ca-f46b-43da-b4e0-80fc31695343'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('0755dcf0-5723-45e4-9b40-e85c4faba68f'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('4d219618-542c-40e4-ba4f-e2e6e4d5d202'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/authenticated/Org1 "HTTP/1.1 200 OK"
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/invited/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('53a7abf7-3d92-42ef-9ee5-f60c7f244ec3'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('12c956a8-57e2-4421-91e1-d86ac4537434'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('f97d1be0-878c-41cc-b848-558b3b2cc186'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/authenticated/Org1 "HTTP/1.1 200 OK"
INFO     parsec.components.postgresql.events:events.py:87 2024-04-29 06:45:56 [info     ] Broadcasting                   [parsec.components.postgresql.events] event_=EventEnrollmentConduit(type='ENROLLMENT_CONDUIT', event_id=UUID('c9d2e513-1895-4299-af50-7dc769ad7159'), organization_id=OrganizationID("Org1"), token=InvitationToken("e0000000000000000000000000000002"), greeter=UserID("alice"))
INFO     httpx:_client.py:1773 HTTP Request: POST http://parsec.invalid/invited/Org1 "HTTP/1.1 200 OK"
```

</details>

--- a/server/parsec/components/invite.py
+++ b/server/parsec/components/invite.py
@@ -77,7 +77,7 @@ class ConduitListenCtx:
     state: ConduitState
     payload: bytes
     peer_payload: bytes | None
-    last_exchange: bool
+    is_last_exchange: bool
 
 
 @dataclass(slots=True)

--- a/server/parsec/components/invite.py
+++ b/server/parsec/components/invite.py
@@ -77,6 +77,7 @@ class ConduitListenCtx:
     state: ConduitState
     payload: bytes
     peer_payload: bytes | None
+    last_exchange: bool
 
 
 @dataclass(slots=True)

--- a/server/parsec/components/memory/invite.py
+++ b/server/parsec/components/memory/invite.py
@@ -170,6 +170,7 @@ class MemoryInviteComponent(BaseInviteComponent):
             state=state,
             payload=payload,
             peer_payload=curr_peer_payload,
+            last_exchange=invitation.conduit_is_last_exchange,
         )
 
     @override
@@ -257,7 +258,12 @@ class MemoryInviteComponent(BaseInviteComponent):
                 invitation.conduit_state == NEXT_CONDUIT_STATE[ctx.state]
                 and curr_our_payload is None
             ):
-                return ctx.peer_payload, invitation.conduit_is_last_exchange
+                # Careful here: it's possible that peer has already changed the state
+                # of the conduit by the time we reach this point. The means it might
+                # have already changed the `last_exchange` flag. We should not rely on.
+                # Instead, we return the one we captured during the `conduit_talk`,
+                # along with the peer payload.
+                return ctx.peer_payload, ctx.last_exchange
 
             elif (
                 invitation.conduit_state != ctx.state

--- a/server/parsec/components/memory/invite.py
+++ b/server/parsec/components/memory/invite.py
@@ -170,7 +170,7 @@ class MemoryInviteComponent(BaseInviteComponent):
             state=state,
             payload=payload,
             peer_payload=curr_peer_payload,
-            last_exchange=invitation.conduit_is_last_exchange,
+            is_last_exchange=invitation.conduit_is_last_exchange,
         )
 
     @override
@@ -247,6 +247,9 @@ class MemoryInviteComponent(BaseInviteComponent):
                     )
                 )
 
+                # The peer payload is the one that was in the base right before we updated the state.
+                # Similarly, the `is_last_exchange` flag is also the one we captured right before updating
+                # the state.
                 return curr_peer_payload, invitation.conduit_is_last_exchange
 
         else:
@@ -258,12 +261,12 @@ class MemoryInviteComponent(BaseInviteComponent):
                 invitation.conduit_state == NEXT_CONDUIT_STATE[ctx.state]
                 and curr_our_payload is None
             ):
-                # Careful here: it's possible that peer has already changed the state
-                # of the conduit by the time we reach this point. The means it might
-                # have already changed the `last_exchange` flag. We should not rely on.
-                # Instead, we return the one we captured during the `conduit_talk`,
-                # along with the peer payload.
-                return ctx.peer_payload, ctx.last_exchange
+                # Careful here: it's possible that the other peer has already sent its payload
+                # for the new state by the time we reach this point. This also means that it might
+                # already have changed the `last_exchange` flag, so we can't rely on what is currently
+                # in the database to determine if this is the last exchange. Instead, we return the
+                # flag we captured during the `conduit_talk` along with the peer payload.
+                return ctx.peer_payload, ctx.is_last_exchange
 
             elif (
                 invitation.conduit_state != ctx.state

--- a/server/parsec/components/postgresql/invite.py
+++ b/server/parsec/components/postgresql/invite.py
@@ -751,6 +751,7 @@ class PGInviteComponent(BaseInviteComponent):
         curr_conduit_state = ConduitState(row["conduit_state"])
         curr_greeter_payload = row["conduit_greeter_payload"]
         curr_claimer_payload = row["conduit_claimer_payload"]
+        curr_last_exchange = row["last_exchange"]
 
         if is_greeter:
             curr_our_payload = curr_greeter_payload
@@ -775,6 +776,7 @@ class PGInviteComponent(BaseInviteComponent):
                 curr_greeter_payload = None
                 curr_our_payload = None
                 curr_peer_payload = None
+                curr_last_exchange = False
             else:
                 return InviteConduitExchangeBadOutcome.ENROLLMENT_WRONG_STATE
 
@@ -790,6 +792,7 @@ class PGInviteComponent(BaseInviteComponent):
                     last=last,
                 )
             )
+            curr_last_exchange = last
         else:
             await conn.execute(
                 *_q_conduit_update(
@@ -817,6 +820,7 @@ class PGInviteComponent(BaseInviteComponent):
             is_greeter=is_greeter,
             payload=payload,
             peer_payload=curr_peer_payload,
+            last_exchange=curr_last_exchange,
         )
 
     @override
@@ -917,7 +921,12 @@ class PGInviteComponent(BaseInviteComponent):
             # has switched to it next state.
 
             if curr_conduit_state == NEXT_CONDUIT_STATE[ctx.state] and curr_our_payload is None:
-                return ctx.peer_payload, is_last_exchange
+                # Careful here: it's possible that peer has already changed the state
+                # of the conduit by the time we reach this point. The means it might
+                # have already changed the `last_exchange` flag. We should not rely on.
+                # Instead, we return the one we captured during the `conduit_talk`,
+                # along with the peer payload.
+                return ctx.peer_payload, ctx.last_exchange
             elif (
                 curr_conduit_state != ctx.state
                 or curr_our_payload != ctx.payload

--- a/server/parsec/components/postgresql/invite.py
+++ b/server/parsec/components/postgresql/invite.py
@@ -820,7 +820,7 @@ class PGInviteComponent(BaseInviteComponent):
             is_greeter=is_greeter,
             payload=payload,
             peer_payload=curr_peer_payload,
-            last_exchange=curr_last_exchange,
+            is_last_exchange=curr_last_exchange,
         )
 
     @override
@@ -913,6 +913,10 @@ class PGInviteComponent(BaseInviteComponent):
                         greeter=ctx.greeter,
                     )
                 )
+
+                # The peer payload is the one that was in the base right before we updated the state.
+                # Similarly, the `is_last_exchange` flag is also the one we captured right before updating
+                # the state.
                 return curr_peer_payload, is_last_exchange
 
         else:
@@ -921,12 +925,12 @@ class PGInviteComponent(BaseInviteComponent):
             # has switched to it next state.
 
             if curr_conduit_state == NEXT_CONDUIT_STATE[ctx.state] and curr_our_payload is None:
-                # Careful here: it's possible that peer has already changed the state
-                # of the conduit by the time we reach this point. The means it might
-                # have already changed the `last_exchange` flag. We should not rely on.
-                # Instead, we return the one we captured during the `conduit_talk`,
-                # along with the peer payload.
-                return ctx.peer_payload, ctx.last_exchange
+                # Careful here: it's possible that the other peer has already sent its payload
+                # for the new state by the time we reach this point. This also means that it might
+                # already have changed the `last_exchange` flag, so we can't rely on what is currently
+                # in the database to determine if this is the last exchange. Instead, we return the
+                # flag we captured during the `conduit_talk` along with the peer payload.
+                return ctx.peer_payload, ctx.is_last_exchange
             elif (
                 curr_conduit_state != ctx.state
                 or curr_our_payload != ctx.payload

--- a/server/tests/api_v4/test_enrollment.py
+++ b/server/tests/api_v4/test_enrollment.py
@@ -1,11 +1,14 @@
 # Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
 
+import asyncio
 import secrets
 
 import anyio
 import pytest
 
 from parsec._parsec import HashDigest, PrivateKey, authenticated_cmds, invited_cmds
+from parsec.components.memory.invite import MemoryInviteComponent
+from parsec.components.postgresql.invite import ConduitState, PGInviteComponent
 from tests.common import CoolorgRpcClients, RpcTransportError
 
 
@@ -17,7 +20,34 @@ from tests.common import CoolorgRpcClients, RpcTransportError
         "user",
     ),
 )
-async def test_device_enrollment(invitation: str, coolorg: CoolorgRpcClients):
+@pytest.mark.parametrize(
+    "issue7212",
+    (
+        False,
+        True,
+    ),
+    ids=("normal", "issue7212"),
+)
+async def test_device_enrollment(
+    invitation: str, coolorg: CoolorgRpcClients, issue7212: bool, monkeypatch
+):
+    if issue7212:
+        pg_conduit_listen = PGInviteComponent._conduit_listen
+        memory_conduit_listen = MemoryInviteComponent._conduit_listen
+
+        async def patched_pg_conduit_listen(self, now, ctx, *args, **kwargs):
+            if not ctx.is_greeter and ctx.state == ConduitState.STATE_4_COMMUNICATE:
+                await asyncio.sleep(0.1)
+            return await pg_conduit_listen(self, now, ctx, *args, **kwargs)
+
+        async def patched_memory_conduit_listen(self, now, ctx, *args, **kwargs):
+            if not ctx.is_greeter and ctx.state == ConduitState.STATE_4_COMMUNICATE:
+                await asyncio.sleep(0.1)
+            return await memory_conduit_listen(self, now, ctx, *args, **kwargs)
+
+        monkeypatch.setattr(PGInviteComponent, "_conduit_listen", patched_pg_conduit_listen)
+        monkeypatch.setattr(MemoryInviteComponent, "_conduit_listen", patched_memory_conduit_listen)
+
     match invitation:
         case "device":
             invitation_token = coolorg.invited_alice_dev3.token

--- a/server/tests/api_v4/test_enrollment.py
+++ b/server/tests/api_v4/test_enrollment.py
@@ -21,32 +21,55 @@ from tests.common import CoolorgRpcClients, RpcTransportError
     ),
 )
 @pytest.mark.parametrize(
-    "issue7212",
+    "delay_state_4_claimer",  # see issue https://github.com/Scille/parsec-cloud/issues/7212
     (
-        False,
-        True,
+        pytest.param(False, id="normal"),
+        pytest.param(True, id="delay_state_4_claimer"),
     ),
-    ids=("normal", "issue7212"),
 )
 async def test_device_enrollment(
-    invitation: str, coolorg: CoolorgRpcClients, issue7212: bool, monkeypatch
+    invitation: str,
+    coolorg: CoolorgRpcClients,
+    delay_state_4_claimer: bool,
+    monkeypatch: pytest.MonkeyPatch,
 ):
-    if issue7212:
-        pg_conduit_listen = PGInviteComponent._conduit_listen
-        memory_conduit_listen = MemoryInviteComponent._conduit_listen
+    if delay_state_4_claimer:
+        # The point here is to reproduce a rare race condition reported in issue #7212 that happens about ~1% of the time on the CI.
+        # The following patch causes it to happen 100% of the time instead.
+        # More precisely, it makes sure the operations happen in the following order:
+        #
+        # 1. State 4A greeter talk, registering `<from greeter 1>` payload
+        # 2. State 4A claimer talk, registering `<from claimer 1>` payload
+        # 3. State 4A greeter listen, updating state to 4B, resetting the payloads and returning `<from claimer 1>`
+        # ** Here 4A claimer listen is delayed for some reason **
+        # 4. State 4B greeter talk, registering `<from greeter 2>` and setting the `last_exchange` flag to `True`
+        # 5. State 4A claimer listen, returning `<from greeter 1>` and `last_exchange` flag to `False`
+        #
+        # Due to a bug, step 5 used to return the `last_exchange` flag to `True` instead of `False`.
 
-        async def patched_pg_conduit_listen(self, now, ctx, *args, **kwargs):
-            if not ctx.is_greeter and ctx.state == ConduitState.STATE_4_COMMUNICATE:
-                await asyncio.sleep(0.1)
-            return await pg_conduit_listen(self, now, ctx, *args, **kwargs)
+        def patch_conduit_listen(cls):
+            greeter_last_listen = asyncio.Event()
+            orignal_method = cls._conduit_listen
 
-        async def patched_memory_conduit_listen(self, now, ctx, *args, **kwargs):
-            if not ctx.is_greeter and ctx.state == ConduitState.STATE_4_COMMUNICATE:
-                await asyncio.sleep(0.1)
-            return await memory_conduit_listen(self, now, ctx, *args, **kwargs)
+            async def patched_conduit_listen(self, now, ctx, *args, **kwargs):
+                if (
+                    not ctx.is_greeter
+                    and ctx.state == ConduitState.STATE_4_COMMUNICATE
+                    and ctx.payload == b"<from claimer 1>"
+                ):
+                    await greeter_last_listen.wait()
+                if (
+                    ctx.is_greeter
+                    and ctx.state == ConduitState.STATE_4_COMMUNICATE
+                    and ctx.payload == b"<from greeter 2>"
+                ):
+                    greeter_last_listen.set()
+                return await orignal_method(self, now, ctx, *args, **kwargs)
 
-        monkeypatch.setattr(PGInviteComponent, "_conduit_listen", patched_pg_conduit_listen)
-        monkeypatch.setattr(MemoryInviteComponent, "_conduit_listen", patched_memory_conduit_listen)
+            monkeypatch.setattr(cls, "_conduit_listen", patched_conduit_listen)
+
+        patch_conduit_listen(PGInviteComponent)
+        patch_conduit_listen(MemoryInviteComponent)
 
     match invitation:
         case "device":

